File: topics/storing-using-data/examples/BikeRace.pas
program GameMain;
uses
  sgTypes, sgInput, sgAudio, sgGraphics, sgResources, sgUtils, sgText;

  // ====================
  // = Define constants =
  // ====================
  const
    WHEEL_SIZE = 10;
    WHEEL_GAP = 10;
    SEAT_GAP = 5;
    MAX_ACCELERATION = 10;
    RACE_DURATION = 30;
    X_SCALE_FACTOR = 0.15;

  // =============
  // = Functions =
  // =============

  // Calculate the distance travelled given acceleration and time
  // distance = ut + ((at^2) / 2)
  function DistanceTravelled(initialSpeed, acceleration, time: Single): Single;
  begin
      result := (initialSpeed * time) + ((acceleration * time * time) / 2);
  end;
  
  // Calculate the x position of a bike accelerating at the given
  // acceleration for the duration of the race 
  function BikeXForAccel(acceleration: Single): Single;
  var
    distance: Single;
  begin
      distance := DistanceTravelled(0, acceleration, RACE_DURATION);
      result := distance * X_SCALE_FACTOR;
  end;

  // Come up with a random acceleration value for a bike between 0 and 
  // MAX_ACCELERATION
  function RandomAccel() : Single;
  begin
      result := rnd() * MAX_ACCELERATION;
  end;

File: topics/storing-using-data/examples/bike-race.c
/* Program: bike-race.c, splashkit */

#include "splashkit.h"

// ====================
// = Define constants =
// ====================

#define WHEEL_SIZE 10
#define WHEEL_GAP 10
#define SEAT_GAP 5
#define MAX_ACCELERATION 10
#define RACE_DURATION 30
#define X_SCALE_FACTOR 0.15

// =============
// = Functions =
// =============

// Calculate the distance travelled given acceleration and time
// distance = ut + ((at^2) / 2)
float distance_travelled(float initial_speed, float acceleration, float time)
{
    return (initial_speed * time) + ((acceleration * time * time) / 2);
}

// Calculate the x position of a bike accelerating at the given
// acceleration for the duration of the race 
float bike_x_for_accel(float acceleration)
{
    float distance;
    
    distance = distance_travelled(0, acceleration, RACE_DURATION);
    
    return distance * X_SCALE_FACTOR;
}

// Come up with a random acceleration value for a bike between 0 and 
// MAX_ACCELERATION
float random_accel()
{
    return rnd() * MAX_ACCELERATION;
}
